### PR TITLE
Add NuGet project update events

### PR DESCRIPTION
**Pull request by nkolev92**
Design for https://github.com/NuGet/Home/issues/9782 primarily and potentially https://github.com/NuGet/Home/issues/6872. 

Hey @lifengl, here's the API we talked about. 

lmk if you have any feedback.

### CODE REVIEW

File: proposed/2021/ProjectUpdateEventsInVisualStudio.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+namespace NuGet.SolutionRestoreManager
+{
+    /// <summary>
+    /// NuGet project update events.
+    /// Architecturally, packages.config and PackageReference projects differ in the way package updates are processed.
+    /// This interface is meant to provide a single API of interest for components wanting to listen to *all* project updates by NuGet.
+    /// </summary>
+    [ComImport]
+    [Guid("30CDDD0A-6901-482D-8CEF-6D798F1A99FC")]
+    public interface IVsNuGetProjectUpdateEvents
```
Comment (lifengl) [inline proposed/2021/ProjectUpdateEventsInVisualStudio.md]: how do we access this component? is the implementation free threaded & do we expect events to be fired in a thread pool thread?  Other than those a few things to clarify. The contract looks good to me.

Comment (nkolev92) [reply proposed/2021/ProjectUpdateEventsInVisualStudio.md]: NuGet exposes services through the NuGet.VisualStudio assembly. You should be able to consume IVsNuGetProjectUpdateEvents via MEF. 

The PackageReference events (only ones you care about), are going to be fired from a threadpool thread. 

I'll need to verify how the packages.config ones are fired today.

File: proposed/2021/ProjectUpdateEventsInVisualStudio.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    {
+        /// <summary>
+        /// Raised when solution restore starts with the list of projects that will be restored.
+        /// The list will not include all projects. Some projects may have been skipped in earlier up to date check, and other projects may no-op.
+        /// </summary>
+        /// <remarks>
+        /// Just because a project is being restored that doesn't necessarily mean any actual updates will happen.
+        /// Only PackageReference projects are includede in this list.
```
Comment (dominoFire) [inline proposed/2021/ProjectUpdateEventsInVisualStudio.md]:
```suggestion
        /// Only PackageReference projects are included in this list.
```

File: proposed/2021/ProjectUpdateEventsInVisualStudio.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        /// </remarks>
+        event SolutionRestoreEventHandler SolutionRestoreStarted;
+
+        /// <summary>
+        /// Raised when solution restore finishes with the list of projects that were restored.
+        /// The list will not include all projects. Some projects may have been skipped in earlier up to date check, and other projects may no-op.
+        /// </summary>
+        /// <remarks>
+        /// Just because a project is being restored that doesn't necessarily mean any actual updates will happen.
+        /// Only PackageReference projects are includede in this list.
```
Comment (dominoFire) [inline proposed/2021/ProjectUpdateEventsInVisualStudio.md]:
```suggestion
        /// Only PackageReference projects are included in this list.
```

File: proposed/2021/ProjectUpdateEventsInVisualStudio.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    /// <param name="projectUniqueName">Project full path. Never <see langword="null"/>. </param>
+    /// <param name="updatedFiles">NuGet output files that may be updated. Never <see langword="null"/>.</param>
+    public delegate void ProjectUpdateEventHandler(string projectUniqueName, IReadOnlyList<string> updatedFiles);
+}
+```
+
+A few other notes:
+
+- `SolutionRestoreStarted` and `SolutionRestoreFinished` events will only be called when a solution restore happens, as a response to a solution/project load, or a user issued restore. Package installation through PM UI, PMC or any of the NuGet extensibility APIs *will never* call this event.
+- `ProjectUpdateStarted` and `ProjectUpdateFinished` may be called without corresponding `SolutionRestoreStarted` and `SolutionRestoreFinished` events, if a package installation has been triggered through PM UI, PMC or any of the NuGet extensibility APIs.
```
Comment (dominoFire) [inline proposed/2021/ProjectUpdateEventsInVisualStudio.md:121]: But PM UI or PMC triggers a restore _after_ a package is modifed (installed, uninstalled, updated)

In that case, will we have 

- ProjectUpdateStarted: From package install
- ProjectUpdateFinished: From package install
- SolutionRestoreStarted: From PMUI/PMC restore action
- SolutionRestoreFinished: From PMUI/PMC restore action

?

**Review by dominoFire — COMMENTED** [review]
Some comments

File: proposed/2021/ProjectUpdateEventsInVisualStudio.md (hunk 4 above)
Comment (nkolev92) [reply proposed/2021/ProjectUpdateEventsInVisualStudio.md:121]: The restore in PackageReference following a package change through PMC or PM UI will be no-op.

**Comment by nkolev92** [comment]
@NuGet/nuget-client 

Hey all, 

The spec is up to date as per our team discussion.

Can someone please approve so I can merge? Thanks!

**Comment by nkolev92** [comment]
:bell @NuGet/nuget-client

**Review by erdembayar — APPROVED** [review]

**Comment by nkolev92** [comment]
Thanks @erdembayar

**Pull request merged**